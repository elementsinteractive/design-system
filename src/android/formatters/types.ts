export type FormatterConfig = {
  template: string
  header: string
  packageName: string
}
