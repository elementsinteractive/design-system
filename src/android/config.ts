import { AndroidConfig } from '../design.types'
import { TEMPLATES } from './constants'
import { AndroidJsonOptions } from './types'

export const createConfig = (input: string, options: AndroidJsonOptions) => {
  const { theme } = options

  const { colorsTemplate, spacingsTemplate, typographyTemplate } =
    TEMPLATES(theme)

  return {
    source: [input],
    platforms: {
      android: {
        transformGroup: 'tokens-android',
        files: [typographyTemplate, spacingsTemplate, colorsTemplate].map(
          template => ({
            format: template.formatter,
            destination: template.destination,
            filter: template.filter,
          }),
        ),
      },
    },
  }
}
