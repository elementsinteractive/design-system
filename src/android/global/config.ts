import { AndroidGlobalOptions } from '../types'
import { GLOBAL_TEMPLATES } from './constants'

export const createGlobalConfig = (
  input: string,
  options: AndroidGlobalOptions,
) => {
  const { spacingsTypeTemplate, colorsTypeTemplate } = GLOBAL_TEMPLATES(options)

  return {
    source: [input],
    platforms: {
      android: {
        transformGroup: 'tokens-android',
        files: [spacingsTypeTemplate, colorsTypeTemplate].map(template => ({
          format: template.formatter,
          destination: template.destination,
          filter: template.filter,
        })),
      },
    },
  }
}
