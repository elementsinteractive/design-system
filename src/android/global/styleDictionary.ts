import StyleDictionaryPackage from 'style-dictionary'

import { useTemplate } from '../../utils'
import { AndroidGlobalOptions } from '../types'
import { GLOBAL_TEMPLATES } from './constants'
import { colorTypeFormatter } from './formatters/colorTypeFormatter'
import { spacingFormatter } from '../theme/formatters/spacingFormatter'

export const globalSDRegisters = (options: AndroidGlobalOptions) => {
  const templateInfo = GLOBAL_TEMPLATES(options)

  const { headerTemplate, spacingsTypeTemplate, colorsTypeTemplate } =
    templateInfo

  const header = useTemplate(headerTemplate)()

  /**
   * Global formatters
   */
  StyleDictionaryPackage.registerFormat({
    name: colorsTypeTemplate.formatter,
    formatter: colorTypeFormatter({
      template: colorsTypeTemplate.source,
      packageName: colorsTypeTemplate.packageName,
      header,
    }),
  })

  StyleDictionaryPackage.registerFormat({
    name: spacingsTypeTemplate.formatter,
    formatter: spacingFormatter(
      {
        template: spacingsTypeTemplate.source,
        packageName: spacingsTypeTemplate.packageName,
        header,
      },
      spacingsTypeTemplate.packageName,
    ),
  })
}
