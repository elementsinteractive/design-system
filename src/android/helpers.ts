import { fileToJSON } from '../common/helpers'

export const readJsonFileAsOptions = <T>(configPath: string) =>
  fileToJSON(configPath) as T
