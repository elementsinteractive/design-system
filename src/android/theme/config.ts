import { THEME_TEMPLATES } from './constants'
import { AndroidThemeOptions } from '../types'

export const createThemeConfig = (
  input: string,
  options: AndroidThemeOptions,
) => {
  const { paletteTemplate, spacingsTemplate, typographyTemplate } =
    THEME_TEMPLATES(options)

  return {
    source: [input],
    platforms: {
      android: {
        transformGroup: 'tokens-android',
        files: [typographyTemplate, spacingsTemplate, paletteTemplate].map(
          template => ({
            format: template.formatter,
            destination: template.destination,
            filter: template.filter,
          }),
        ),
      },
    },
  }
}
