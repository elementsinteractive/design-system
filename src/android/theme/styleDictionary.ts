import StyleDictionaryPackage from 'style-dictionary'

import { useTemplate } from '../../utils'
import { AndroidThemeOptions } from '../types'

import { THEME_TEMPLATES } from './constants'
import { spacingFormatter } from './formatters/spacingFormatter'
import { typographyFormatter } from './formatters/typographyFormatter'
import { paletteFormatter } from './formatters/paletteFormatter'
import { colorFormatter } from './formatters/colorFormatter'

export const themeSDRegisters = (options: AndroidThemeOptions) => {
  const templateInfo = THEME_TEMPLATES(options)

  const {
    headerTemplate,
    paletteTemplate,
    spacingsTemplate,
    typographyTemplate,
    colorTokensTemplate,
  } = templateInfo

  const header = useTemplate(headerTemplate)()

  /**
   * Formatters
   */
  StyleDictionaryPackage.registerFormat({
    name: colorTokensTemplate.formatter,
    formatter: colorFormatter(
      {
        template: colorTokensTemplate.source,
        header,
        packageName: colorTokensTemplate.packageName,
      },
      options.globalPackageName,
      options.theme.name,
    ),
  })

  StyleDictionaryPackage.registerFormat({
    name: typographyTemplate.formatter,
    formatter: typographyFormatter({
      template: typographyTemplate.source,
      header,
      packageName: typographyTemplate.packageName,
    }),
  })

  StyleDictionaryPackage.registerFormat({
    name: spacingsTemplate.formatter,
    formatter: spacingFormatter(
      {
        template: spacingsTemplate.source,
        header,
        packageName: spacingsTemplate.packageName,
      },
      spacingsTemplate.sharedPackageName,
    ),
  })

  StyleDictionaryPackage.registerFormat({
    name: paletteTemplate.formatter,
    formatter: paletteFormatter({
      template: paletteTemplate.source,
      header,
      packageName: paletteTemplate.packageName,
    }),
  })

  /**
   * Filters
   */
  const registerFilter = (name, type) => {
    StyleDictionaryPackage.registerFilter({
      name,
      matcher: token => token.type === type,
    })
  }

  registerFilter(typographyTemplate.filter, 'typography')
  registerFilter(spacingsTemplate.filter, 'spacing')
  registerFilter(paletteTemplate.filter, 'color')

  StyleDictionaryPackage.registerTransformGroup({
    name: 'tokens-android',
    transforms: ['name/ti/camel'],
  })
}
