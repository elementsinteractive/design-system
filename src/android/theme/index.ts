import StyleDictionaryPackage from 'style-dictionary'

import { readJsonFileAsOptions } from '../helpers'
import { AndroidThemeOptions } from '../types'
import { themeSDRegisters } from './styleDictionary'
import { createThemeConfig } from './config'

export const getThemeSD = (configPath: string, input: string) => {
  const options = readJsonFileAsOptions<AndroidThemeOptions>(configPath)

  console.log('Retrieving theme SD using options', options)

  themeSDRegisters(options)

  return StyleDictionaryPackage.extend(createThemeConfig(input, options))
}
