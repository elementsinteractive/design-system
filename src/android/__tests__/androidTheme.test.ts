import StyleDictionaryPackage from 'style-dictionary'
import { clearOutput, fileExists, fileToString } from '../../common/helpers'
import { createThemeConfig } from '../theme/config'
import {
  PaletteFilename,
  SpacingFilename,
  TypeFilename,
} from '../theme/constants'
import { readJsonFileAsOptions } from '../helpers'
import { AndroidThemeOptions } from '../types'
import { themeSDRegisters } from '../theme/styleDictionary'

describe('android', () => {
  var config = {
    configPath: 'src/android/__tests__/config.json',
    input: 'src/__tests__/tokens.json',
  }

  let StyleDictionaryExtended

  let options: AndroidThemeOptions

  beforeAll(() => {
    const { input, configPath } = config

    options = readJsonFileAsOptions(configPath)

    themeSDRegisters(options)

    StyleDictionaryExtended = StyleDictionaryPackage.extend(
      createThemeConfig(input, options),
    )
  })

  beforeEach(() => {
    clearOutput('src/android/__tests__/__output')
  })

  it('can setup with global config', () => {
    StyleDictionaryExtended.buildPlatform('android')

    expect(
      StyleDictionaryExtended.transformGroup['tokens-android'],
    ).toMatchSnapshot()

    const colorFile = `${options.theme.colors.destination}/${PaletteFilename}`
    const spacingFile = `${options.theme.spacings.destination}/${SpacingFilename}`
    const typeFile = `${options.theme.typography.destination}/${TypeFilename}`

    const files = [colorFile, spacingFile, typeFile]

    files.forEach(path => {
      expect(fileExists(path)).toBeTruthy()
      expect(fileToString(path)).toMatchSnapshot()
    })
  })
})
