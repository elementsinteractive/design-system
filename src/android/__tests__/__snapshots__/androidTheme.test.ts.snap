// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`android can setup with global config 1`] = `
Array [
  "name/ti/camel",
]
`;

exports[`android can setup with global config 2`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package org.example.design.base

import androidx.compose.ui.graphics.Color

val primaryDefault = Color(0xffffffff)

val primaryBackground = Color(0xfff4f4f4)

val primaryVariant = Color(0xff000000)

val secondary = Color(0xffff495b)

val secondaryVariant = Color(0xfff4f4f4)

val surface = Color(0xffffffff)

val onPrimary = Color(0xffffffff)

val onSecondary = Color(0xff000000)

val onBackground = Color(0xff000000)

val onSurface = Color(0xff000000)

val purple = Color(0xff5048da)

val white = Color(0xffffffff)

val pink = Color(0xffff495b)

val black = Color(0xff000000)

val lightGray = Color(0xfff4f4f4)

"
`;

exports[`android can setup with global config 3`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package org.example.design.base

import .SpacingTokens

object Spacings : SpacingTokens {
  override val xxxs = 4
  override val xxs = 8
  override val xs = 12
  override val s = 16
  override val m = 24
  override val l = 32
  override val xl = 48
  override val xxl = 200
}"
`;

exports[`android can setup with global config 4`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package org.example.design.base

import androidx.compose.material.Typography
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp

val h1 = TextStyle(
    fontWeight = FontWeight.Light,
    fontSize = 96.sp,
    letterSpacing = (-1.5).sp,
)

val h2 = TextStyle(
    fontWeight = FontWeight.Light,
    fontSize = 60.sp,
    letterSpacing = (-0.5).sp,
)

val h3 = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 48.sp,
)

val h4 = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 34.sp,
    letterSpacing = (0.25).sp,
)

val h5 = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 24.sp,
)

val h6 = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 20.sp,
    letterSpacing = (0.15).sp,
)

val body1 = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 16.sp,
    letterSpacing = (0.5).sp,
)

val body2 = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 14.sp,
    letterSpacing = (0.25).sp,
)

val subtitle1 = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 16.sp,
    letterSpacing = (0.15).sp,
)

val subtitle2 = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 14.sp,
    letterSpacing = (0.1).sp,
)

val button = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 14.sp,
    letterSpacing = (1.25).sp,
)

val caption = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 12.sp,
    letterSpacing = (0.4).sp,
)

val overline = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 10.sp,
    letterSpacing = (1.5).sp,
)

val AppTypography = Typography(
    h1 = h1,
    h2 = h2,
    h3 = h3,
    h4 = h4,
    h5 = h5,
    h6 = h6,
    body1 = body1,
    body2 = body2,
    subtitle1 = subtitle1,
    subtitle2 = subtitle2,
    button = button,
    caption = caption,
    overline = overline,
)"
`;
